### PR TITLE
test: Break if DEBUG_ASSERT is encountered

### PR DESCRIPTION
**Pull request by Holzhaus**
**TODO:**
- [x] Fix ControlObject/ControlPrivateDouble issues #2918
- [x] Fix debug assertions in TagLibTest #2912 
- [x] Fix debug assertions in AnalyzerWaveformTest #2927
- [x] Fix debug assertion in SoundSourceFFmpeg #2931
- [x] Fix debug assertions in AutoDJProcessorTest ~~#2929~~ #2951 
- [x] ~~Add missing COs to MixxxTest~~ Fix CO debug assertions caused by the PlayerInfo timer #2988
- [x] Fix debug assertion for SoundSourceMediaFoundation (Windows) #3002 

If you build Mixxx in debug mode, this makes sure that debug assertions lead to test failures. This implementation is far from perfect, it does not work in release builds (since debug assertions are stripped in that case), but it would require some bigger changes to make it possible to enable debug assertions for tests but not mixxx itself. For now, always make sure to build tests in debug mode or with DEBUG_ASSERTS_FATAL=ON.

**Note:** This will lead to a bunch of test failures. Not sure if we want to merge this before we fix them, because it will lead to failed CI builds in unrelated branches.

### CODE REVIEW

**Comment by daschuer** [comment]
I have not yet understand how logging relates the DEBUG_ASSERT. 

But let's release all from headaches and hold this back until the all test are working.

**Comment by uklotzde** [comment]
#2912

**Comment by daschuer** [comment]
> I grepped for MIXXX_DEBUG_ASSERTIONS_FATAL and discovered it's enabled by installing a Qt message handler 

It sounds like his should be decoupled. Right?

**Comment by Holzhaus** [comment]
> > I grepped for MIXXX_DEBUG_ASSERTIONS_FATAL and discovered it's enabled by installing a Qt message handler
> 
> It sounds like his should be decoupled. Right?

Not necessarily. The elephant in the room is that tests will only work reliably (as in: they fail when shit hits the fan) when the `MIXXX_BUILD_DEBUG` preprocessor definition is set when compiling Mixxx. Otherwise `DEBUG_ASSERT` statements are stripped from the source, so tests will pass even though they do something nasty. That means we always need to build in debug mode during development.

The only way around this that I can think of would be to compile the `mixxx-lib` cmake target twice, one for the `mixxx` executable (possibly without debug asssertions) and one for the tests (always with debug assertions). I don't think that's feasible.

Let's always do debug builds on the CI for PRs and testing our branches, and we can add a second build stage that is only executed when operating on a tag (not a PR/branch) and builds in release mode for deployment.

**Comment by Holzhaus** [comment]
My last commit should make CI builds fail. When `DEBUG_ASSERTIONS_FATAL` has been set, we should enable `DEBUG_ASSERT` statements even in Release builds. For CMake I added a status message in that case.

For CI testing we also just switch to debug builds, but I think if a user explicitly enabled the `DEBUG_ASSERTIONS_FATAL` flag, it should be honored independent of the build type IMHO.

**Comment by uklotzde** [comment]
I will try to fix some issues in ControlObject/ControlPrivateDouble that seem to affect us now. Only the bare minimum, no more.

**Comment by Holzhaus** [comment]
> I will try to fix some issues in ControlObject/ControlPrivateDouble that seem to affect us now. Only the bare minimum, no more.

Can you elaborate what issues you're referring to?

**Comment by uklotzde** [comment]
#2918 Issues like leaking controls that are not removed from the internal, static hash map after finishing tests and that are causing subsequent tests to fail with debug assertions when registering new ControlObject instances.

**Comment by Holzhaus** [comment]
I rebased to update the test failure logs (this PR is so small that I don't think rebasing is an issue here).

**Comment by uklotzde** [comment]
> I rebased to update the test failure logs (this PR is so small that I don't think rebasing is an issue here).

Yes, please always rebase after prerequisite PRs have been merged. This PR should remain the icing on the cake, keeping track of the those changes until the tests finally succeed.

**Comment by uklotzde** [comment]
@daschuer Ping. We need your support for fixing the AutoDJ assertions and test failures.

**Comment by uklotzde** [comment]
Please open a separate PR for the /lib integration works.

**Comment by Holzhaus** [comment]
I rebased on latest 2.3, all tests pass now both with ctest and when running `mixxx-test` directly. Merge?

**Comment by uklotzde** [comment]
Please rebase once again.

We might also need to reduce the log level. The CI builds now fail because of the huge amount of debug log messages in some tests.

**Comment by Holzhaus** [comment]
Done. CTest hides the test output and only shows it if a test failed. So if the build is terminated due to the long output, that should be a side effect of the failing test, not the cause.

**Comment by uklotzde** [comment]
> Done. CTest hides the test output and only shows it if a test failed. So if the build is terminated due to the long output, that should be a side effect of the failing test, not the cause.

I don't think so. The GlobalTrackCache test is aborted while running without any failures:
```
Debug []: GlobalTrackCache - Cache miss - allocating track ["/Users/travis/build/mixxxdj/mixxx/src/test/id3-test-data/cover-test.flac" | "/Users/travis/build/mixxxdj/mixxx/src/test/id3-test-data/cover-test.flac" | -1]


Debug []: GlobalTrackCa

The job exceeded the maximum log length, and has been terminated.
```

**Comment by uklotzde** [comment]
The log output might not be shown, but it seems to be written in any case.

**Comment by uklotzde** [comment]
Change log level from Debug to Info during tests:
https://github.com/uklotzde/mixxx/commit/eb76a9b86a5770dcf4df9c151e902f90e18d3708

**Comment by Holzhaus** [comment]
Hmm I'm still not sure if the log level messages is the actual cause of the issue but if you checked it then it probably is. However, I'd rather not disable debug messages for tests globally. Maybe we can add a flag or set it via an env var, so that you can still see debug messages when running locally. I'll poke around tomorrow.

**Comment by Holzhaus** [comment]
For the macOS/CMake build, i think the test times out:

`271/616 Test #271: GlobalTrackCacheTest.concurrentDelete ...........................................***Timeout  30.03 sec`

See Line 4090: https://travis-ci.org/github/mixxxdj/mixxx/jobs/716093492#L4090

A timeout is considered a failure, which is why ctest then starts to print the logging output (which Travis cuts off). But the actual issue seems to be the timeout after 30 seconds.

For SCons, the build still fails due to the verbose log, so changing the test log level is needed anyway as long as we keep SCons. When we drop it, it'd prefer to go back to debug tbh.

File: src/util/cmdlineargs.h [hunk 1]
```diff
@@ -11,6 +11,8 @@
 // A structure to store the parsed command-line arguments
 class CmdlineArgs final {
   public:
+    CmdlineArgs();
```
Comment (uklotzde) [inline src/util/cmdlineargs.h:14]: Please add a remark that the default constructor is only supposed to be used in tests and that all operational code must access the singleton via Instance().

**Comment by uklotzde** [comment]
Thanks for customizing the log level during tests! Please fix the remaining formatting issues.

Detailed logging is not particularly important for the CI builds which are only supposed to detect test failures. In case of any failures, you need to repeat the affected tests anyway. Test failures should be reproducible and you could enable more detailed logging on demand when analyzing failing tests.

**Comment by uklotzde** [comment]
The time out for GlobalTrackCacheTest.concurrentDelete is tight and it fails infrequently on the poor macOS VMs on Travis. Usually, it finishes in about ~25 sec. Unrelated to the changes in this PR.

**Comment by uklotzde** [comment]
These changes greatly improve the confidence in our testing efforts and finally enable a huge amount of additional checks to detect issues early. Thank you!

LGTM

**Comment by uklotzde** [comment]
Master builds may now fail until https://github.com/mixxxdj/mixxx/pull/3006 has been merged.

**Pull request merged**